node/metered-channel/src/tests.rs

```diff
@@ -0,0 +1,129 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use super::*;
+use assert_matches::assert_matches;
+use futures::{executor::block_on, StreamExt};
+
+#[derive(Clone, Copy, Debug, Default)]
+struct Msg {
+	val: u8,
+}
+
+#[test]
+fn try_send_try_next() {
+	block_on(async move {
+		let (mut tx, mut rx) = channel::<Msg>(5);
+		let msg = Msg::default();
+		assert_matches!(rx.meter().read(), Readout { sent: 0, received: 0, .. });
+		tx.try_send(msg).unwrap();
+		assert_matches!(tx.meter().read(), Readout { sent: 1, received: 0, .. });
+		tx.try_send(msg).unwrap();
+		tx.try_send(msg).unwrap();
+		tx.try_send(msg).unwrap();
+		assert_matches!(tx.meter().read(), Readout { sent: 4, received: 0, .. });
+		rx.try_next().unwrap();
+		assert_matches!(rx.meter().read(), Readout { sent: 4, received: 1, .. });
+		rx.try_next().unwrap();
+		rx.try_next().unwrap();
+		assert_matches!(tx.meter().read(), Readout { sent: 4, received: 3, tof } => {
+			// every second in test, consumed before
+			assert_eq!(dbg!(tof).len(), 1);
+		});
+		rx.try_next().unwrap();
+		assert_matches!(rx.meter().read(), Readout { sent: 4, received: 4, tof } => {
+			// every second in test, consumed before
+			assert_eq!(dbg!(tof).len(), 0);
+		});
+		assert!(rx.try_next().is_err());
+	});
+}
+
+#[test]
+fn with_tasks() {
+	let (ready, go) = futures::channel::oneshot::channel();
+
+	let (mut tx, mut rx) = channel::<Msg>(5);
+	block_on(async move {
+		futures::join!(
+			async move {
+				let msg = Msg::default();
+				assert_matches!(tx.meter().read(), Readout { sent: 0, received: 0, .. });
+				tx.try_send(msg).unwrap();
+				assert_matches!(tx.meter().read(), Readout { sent: 1, received: 0, .. });
+				tx.try_send(msg).unwrap();
+				tx.try_send(msg).unwrap();
+				tx.try_send(msg).unwrap();
+				ready.send(()).expect("Helper oneshot channel must work. qed");
+			},
+			async move {
+				go.await.expect("Helper oneshot channel must work. qed");
+				assert_matches!(rx.meter().read(), Readout { sent: 4, received: 0, .. });
+				rx.try_next().unwrap();
+				assert_matches!(rx.meter().read(), Readout { sent: 4, received: 1, .. });
+				rx.try_next().unwrap();
+				rx.try_next().unwrap();
+				assert_matches!(rx.meter().read(), Readout { sent: 4, received: 3, .. });
+				rx.try_next().unwrap();
+				assert_matches!(dbg!(rx.meter().read()), Readout { sent: 4, received: 4, .. });
+			}
+		)
+	});
+}
+
+use futures_timer::Delay;
+use std::time::Duration;
+
+#[test]
+fn stream_and_sink() {
```

Comment:
I don't understand what this test does. Is it just meant to ensure that we don't dead-lock on the write and read of the channels

Reply:
It's meant as a basic smoke test.